crates/vm/src/system/memory/controller/memory.rs

```diff
@@ -406,17 +487,148 @@ impl<F: PrimeField32> Memory<F> {
+            _marker: Default::default(),
+        }
+    }
+
+    // just for unit testing
+    #[cfg(test)]
+    fn last_record(&self) -> &MemoryRecord<F> {
+        self.log.last().unwrap().as_ref().unwrap()
+    }
+}
+
+pub type MemoryImage<F> = FxHashMap<Address, F>;
+
+/// A partition of data into blocks where each block has size a power of two.
+#[derive(Debug, Clone)]
+pub struct Memory<F> {
+    pub(super) data: FxHashMap<Address, F>,
+    pub(super) log: Vec<MemoryLogEntry<F>>,
+    timestamp: u32,
+}
+
+impl<F: PrimeField32> Default for Memory<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<F: PrimeField32> Memory<F> {
+    pub fn new() -> Self {
+        Self {
+            data: MemoryImage::default(),
+            timestamp: INITIAL_TIMESTAMP + 1,
+            log: vec![],
+        }
+    }
+
+    /// Instantiates a new `Memory` data structure from an image.
+    pub fn from_image(image: MemoryImage<F>) -> Self {
+        Self {
+            data: image,
+            timestamp: INITIAL_TIMESTAMP + 1,
+            log: vec![],
+        }
+    }
+
+    fn last_record_id(&self) -> RecordId {
+        RecordId(self.log.len() - 1)
+    }
+
+    /// Writes an array of values to the memory at the specified address space and start index.
+    ///
+    /// Returns the `RecordId` for the memory record and the previous data.
+    pub fn write<const N: usize>(
+        &mut self,
+        address_space: u32,
+        pointer: u32,
+        values: [F; N],
+    ) -> (RecordId, [F; N]) {
+        assert!(N.is_power_of_two());
+
+        let prev_data = array::from_fn(|i| {
+            self.data
+                .insert((address_space, pointer + i as u32), values[i])
+                .unwrap_or(F::ZERO)
+        });
+
+        self.log.push(MemoryLogEntry::Write {
+            address_space,
+            pointer,
+            data: values.to_vec(),
+        });
+        self.timestamp += 1;
+
+        (self.last_record_id(), prev_data)
+    }
+
+    /// Reads an array of values from the memory at the specified address space and start index.
+    pub fn read<const N: usize>(&mut self, address_space: u32, pointer: u32) -> (RecordId, [F; N]) {
+        assert!(N.is_power_of_two());
+
+        self.log.push(MemoryLogEntry::Read {
+            address_space,
+            pointer,
+            len: N,
+        });
+
+        let values = if address_space == 0 {
+            assert_eq!(N, 1, "cannot batch read from address space 0");
+            [F::from_canonical_u32(pointer); N]
+        } else {
+            self.range_array::<N>(address_space, pointer)
+        };
+        self.timestamp += 1;
+        (self.last_record_id(), values)
+    }
+
+    pub fn increment_timestamp_by(&mut self, amount: u32) {
+        self.timestamp += amount;
+        self.log.push(MemoryLogEntry::IncrementTimestampBy(amount))
+    }
+
+    pub fn increase_timestamp_to(&mut self, amount: u32) {
```

Comment:
note to self to look into keccak; as discussed, would be nicer if we didn't need timestamp in online memory

Reply:
I removed this function now.